sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/rule/JoinReorderingTest.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.impl.rule;
+
+import java.math.BigInteger;
+import org.apache.beam.sdk.extensions.sql.impl.BeamSqlEnv;
+import org.apache.beam.sdk.extensions.sql.impl.rel.BeamIOSourceRel;
+import org.apache.beam.sdk.extensions.sql.impl.rel.BeamRelNode;
+import org.apache.beam.sdk.extensions.sql.meta.provider.test.TestTableProvider;
+import org.apache.beam.sdk.values.Row;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * This test ensures that we are reordering joins and get a plan similar to Join(large,Join(small,
+ * medium)) instead of Join(small, Join(medium,large).
+ */
+public class JoinReorderingTest {
+
+  @Test
+  public void test_table_sizes() {
+    TestTableProvider tableProvider = new TestTableProvider();
+    BeamSqlEnv env = BeamSqlEnv.withTableProvider(tableProvider);
+    createThreeTables(env, tableProvider);
+    Assert.assertEquals(
+        BigInteger.ONE,
+        tableProvider
+            .buildBeamSqlTable(tableProvider.getTable("small_table"))
+            .getRowCount(null)
+            .getRowCount());
+
+    Assert.assertEquals(
+        BigInteger.valueOf(3),
+        tableProvider
+            .buildBeamSqlTable(tableProvider.getTable("medium_table"))
+            .getRowCount(null)
+            .getRowCount());
+
+    Assert.assertEquals(
+        BigInteger.valueOf(100),
+        tableProvider
+            .buildBeamSqlTable(tableProvider.getTable("large_table"))
+            .getRowCount(null)
+            .getRowCount());
+  }
+
+  @Test
+  public void test_correct_reordering() {
+    TestTableProvider tableProvider = new TestTableProvider();
+    BeamSqlEnv env = BeamSqlEnv.withTableProvider(tableProvider);
+    createThreeTables(env, tableProvider);
+
+    BeamRelNode parsedQuery =
+        env.parseQuery(
+            "select * from large_table "
+                + " JOIN medium_table on large_table.medium_key = medium_table.large_key "
+                + " JOIN small_table on medium_table.small_key = small_table.medium_key ");
+    assertLargeIsTop(parsedQuery);
+
+    parsedQuery =
```

Comment:
It's better to have many small tests, the goal is to be able to identify the problem quickly looking at the failure. I would split this into tests like `testSmallTableOnTheLeft()`, `testSmallTableOnTheRight()` etc